### PR TITLE
Add Avalara endpoints

### PR DESCRIPTION
**Pull request by subenson**
## Checklist

- [x] Writing style
- [x] API design standards

### CODE REVIEW

File: openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+      - apiToken
+    properties:
+      apiToken:
+        type: string
+        description: Avalara API token.
+        format: password
+        example: 0ab15e3c1cc608b6f592fbddaddeb8f2
+      _links:
+        type: array
+        description: Related resource links.
```
Comment (noon-dawg) [inline openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml:15]: nitpick: suggestion to help make the description more concise
```suggestion
        description: Related links.
```

File: openapi/components/schemas/Invoices/TaxTypes/rebilly-avalara.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+description: Avalara-based taxes.
+allOf:
+  - $ref: ../InvoiceTax.yaml
+  - type: object
+    properties:
+      items:
+        type: array
+        description: |-
+          Taxes that are applied to this invoice and calculated based
+          on the related product's `taxCategoryId`.
```
Comment (noon-dawg) [inline openapi/components/schemas/Invoices/TaxTypes/rebilly-avalara.yaml:10]: issue: [Possessive apostrophes](https://github.com/Rebilly/api-definitions/blob/main/WRITING-STYLE.md#possessive-apostrophes)


```suggestion
          on the related product `taxCategoryId`.
```

File: openapi/components/schemas/Invoices/InvoiceTax.yaml [hunk 3]
```diff
@@ -7,13 +7,15 @@ discriminator:
   mapping:
     manual: ./TaxTypes/manual.yaml
     rebilly-taxjar: ./TaxTypes/rebilly-taxjar.yaml
+    rebilly-avalara: ./TaxTypes/rebilly-avalara.yaml
```
Comment (adamaltman) [inline openapi/components/schemas/Invoices/InvoiceTax.yaml:10]: This file is incorrect. You didn't break it. But, the object that defines the discriminator must use anyOf, oneOf, or allOf and the discriminated property cannot be defined inline.

https://redocly.com/docs/resources/discriminator/

**Review by adamaltman — DISMISSED** [review]

File: openapi/openapi.yaml [hunk 4]
```diff
@@ -588,6 +588,12 @@ tags:
       transactions for the Sales Tax AutoFile feature in your TaxJar
       account. For more information, see
       [TaxJar integration](https://www.rebilly.com/docs/automations/tax-jar-integration/#taxjar-integration).
+  - name: Avalara credentials
+    description: |-
+      Use these operations to manage your Avalara account credentials.
+      The Avalara integration enables you to calculate taxes and create
+      transactions in your Avalara account. For more information, see
```
Comment (adamaltman) [inline openapi/openapi.yaml:595]: https://sembr.org/
```suggestion
      The Avalara integration enables you to calculate taxes and create transactions in your Avalara account.
      For more information, see
```

File: openapi/paths/credential-hashes@avalara.yaml [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    '401':
+      $ref: ../components/responses/Unauthorized.yaml
+    '403':
+      $ref: ../components/responses/Forbidden.yaml
+post:
+  x-products:
+    - Users
+  tags:
+    - Avalara credentials
+  summary: Create a Avalara credential
```
Comment (adamaltman) [inline openapi/paths/credential-hashes@avalara.yaml:35]:
```suggestion
  summary: Create an Avalara credential
```

File: openapi/paths/credential-hashes@avalara.yaml [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+  operationId: PostAvalaraCredentialHash
+  x-sdk-operation-name: createAvalaraCredential
+  description: |-
+    Creates a Avalara credential.
```
Comment (adamaltman) [inline openapi/paths/credential-hashes@avalara.yaml:39]:
```suggestion
    Creates an Avalara credential.
```

Comment (adamaltman) [inline openapi/paths/credential-hashes@avalara.yaml:39]:
```suggestion
  description: Creates an Avalara credential.
```

File: openapi/paths/credential-hashes@avalara@{hash}.yaml [hunk 7]
```diff
@@ -0,0 +1,57 @@
+parameters:
+  - $ref: ../components/parameters/hash.yaml
+get:
+  x-products:
+    - Users
+  tags:
+    - Avalara credentials
+  summary: Retrieve a Avalara credential
```
Comment (adamaltman) [inline openapi/paths/credential-hashes@avalara@{hash}.yaml:8]:
```suggestion
  summary: Retrieve an Avalara credential
```

File: openapi/paths/credential-hashes@avalara@{hash}.yaml [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  operationId: GetAvalaraCredentialHash
+  x-sdk-operation-name: getAvalaraCredential
+  description: |-
+    Retrieves a Avalara credential with a specified token ID.
```
Comment (adamaltman) [inline openapi/paths/credential-hashes@avalara@{hash}.yaml:12]:
```suggestion
  description: Retrieves an Avalara credential with a specified token ID.
```

File: openapi/paths/credential-hashes@avalara@{hash}.yaml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+patch:
+  x-products:
+    - Users
+  tags:
+    - Avalara credentials
+  summary: Update a Avalara credential
+  operationId: PatchAvalaraCredentialHash
+  x-sdk-operation-name: updateAvalaraCredential
+  description: |-
+    Updates a Avalara credential with a specified token ID.
```
Comment (adamaltman) [inline openapi/paths/credential-hashes@avalara@{hash}.yaml:35]:
```suggestion
  description: Updates an Avalara credential with a specified token ID.
```

File: openapi/paths/credential-hashes@avalara@{hash}.yaml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    '403':
+      $ref: ../components/responses/Forbidden.yaml
+    '404':
+      $ref: ../components/responses/NotFound.yaml
+patch:
+  x-products:
+    - Users
+  tags:
+    - Avalara credentials
+  summary: Update a Avalara credential
```
Comment (adamaltman) [inline openapi/paths/credential-hashes@avalara@{hash}.yaml:31]:
```suggestion
  summary: Update an Avalara credential
```

File: openapi/components/schemas/Invoices/InvoiceTax.yaml (hunk 3 above)
Comment (subenson) [reply openapi/components/schemas/Invoices/InvoiceTax.yaml:10]: I added anyOf and removed the inline definition. Thank you.

**Review by adamaltman — DISMISSED** [review]

File: plugins/products-bundler/mapping/Combined.yaml [hunk 11]
```diff
@@ -89,6 +89,7 @@ x-tagGroups:
       - Plaid credentials
       - Experian credentials
       - TaxJar credentials
+      - Avalara credentials
```
Comment (adamaltman) [inline plugins/products-bundler/mapping/Combined.yaml:92]: Not sure about the positioning in this list. Should this list be alphabetized?

Comment (subenson) [reply plugins/products-bundler/mapping/Combined.yaml:92]: I think that makes more sense.

Comment (subenson) [reply plugins/products-bundler/mapping/Combined.yaml:92]: Actually the list itself is a little bit meaningless to me. I'm not sure where it's used, but can we extract the credentials and create a new list for it?

Comment (subenson) [reply plugins/products-bundler/mapping/Combined.yaml:92]: https://github.com/Rebilly/api-definitions/pull/1118/files

**Comment by subenson** [comment]
@Rebilly/js @Rebilly/php please review

File: openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml [hunk 12]
```diff
@@ -0,0 +1,19 @@
+allOf:
+  - $ref: ../Credential.yaml
+  - type: object
+    description: Avalara credential.
+    required:
+      - apiToken
+    properties:
+      apiToken:
```
Comment (ashkarpetin) [inline openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml:8]: It seems incorrect? Shouldn't it be licenseKey and accountId?

File: openapi/components/schemas/Invoices/TaxTypes/rebilly-avalara.yaml [hunk 13]
```diff
@@ -0,0 +1,19 @@
+description: Avalara-based taxes.
```
Comment (adamaltman) [inline openapi/components/schemas/Invoices/TaxTypes/rebilly-avalara.yaml:1]: This isn't good. It has no type declaration and then pairs with an `allOf`. Remove it.
```suggestion
```

File: openapi/paths/credential-hashes@avalara.yaml [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+    - $ref: ../components/parameters/collectionOffset.yaml
+    - $ref: ../components/parameters/collectionSort.yaml
+    - $ref: ../components/parameters/collectionQuery.yaml
+  tags:
+    - Avalara credentials
+  summary: Retrieve Avalara credentials
+  operationId: GetAvalaraCredentialHashCollection
+  x-sdk-operation-name: getAllAvalaraCredentials
+  description: |-
+    Retrieves a list of Avalara credentials.
```
Comment (adamaltman) [inline openapi/paths/credential-hashes@avalara.yaml:16]:
```suggestion
  description: Retrieves a list of Avalara credentials.
```
The purpose of the block scalar is multi-line descriptions.

File: openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml (hunk 12 above)
Comment (subenson) [reply openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml]: Yes it's incorrect

**Comment by subenson** [comment]
@php please review

File: openapi/components/schemas/Invoices/TaxTypes/rebilly-avalara.yaml [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+      - calculator
+    properties:
+      calculator:
+        description: Type of tax calculator.
+        type: string
+      items:
+        type: array
+        description: |-
+          Taxes that are applied to this invoice and calculated based
+          on the related product `taxCategoryId`.
```
Comment (subenson) [inline openapi/components/schemas/Invoices/TaxTypes/rebilly-avalara.yaml:14]:
```suggestion
        description: Taxes that are applied to this invoice and calculated based on the related product `taxCategoryId`.
```

**Comment by subenson** [comment]
@Rebilly/php @Rebilly/docs please review

**Review by ashkarpetin — DISMISSED** [review]

File: openapi/openapi.yaml [hunk 16]
```diff
@@ -510,6 +510,11 @@ tags:
       Use these operations to manage your TaxJar account credentials.
       The TaxJar integration enables you to calculate taxes and create transactions for the Sales Tax AutoFile feature in your TaxJar account.
       For more information, see [TaxJar integration](https://www.rebilly.com/docs/automations/tax-jar-integration/#taxjar-integration).
+  - name: Avalara credentials
+    description: |-
+      Use these operations to manage your Avalara account credentials.
+      The Avalara integration enables you to calculate taxes and create transactions in your Avalara account.
+      For more information, see [Avalara integration](https://www.rebilly.com/docs/automations/avalara-integration/#avalara-integration).
```
Comment (noon-dawg) [inline openapi/openapi.yaml:517]: issue: we don't have these docs yet

I think we should remove this line for now and add the link when we have them

When we create these docs, I think we should add them to `/docs/app-store/` and not `/docs/automations/` ([see the App Store for an example](https://www.rebilly.com/docs/app-store/app-store/))
```suggestion
```

File: openapi/components/schemas/Invoices/TaxTypes/rebilly-taxjar.yaml [hunk 17]
```diff
@@ -1,12 +1,15 @@
...
+      - calculator
     properties:
+      calculator:
+        description: Type of tax calculator.
+        type: string
       items:
         type: array
-        description: |-
-          Taxes that are applied to this invoice and calculated based on the related product's `taxCategoryId`.
+        description: Taxes that are applied to this invoice and calculated based on the related product's `taxCategoryId`.
```
Comment (noon-dawg) [inline openapi/components/schemas/Invoices/TaxTypes/rebilly-taxjar.yaml:12]: issue: where possible, avoid [possessive apostrophes](https://github.com/Rebilly/api-definitions/blob/main/WRITING-STYLE.md#possessive-apostrophes)
```suggestion
        description: Taxes that are applied to this invoice and calculated based on the `taxCategoryId` of the related product .
```

File: openapi/components/schemas/Invoices/TaxTypes/rebilly-avalara.yaml [hunk 18]
```diff
@@ -0,0 +1,16 @@
...
+  - type: object
+    required:
+      - calculator
+    properties:
+      calculator:
+        description: Type of tax calculator.
+        type: string
+      items:
+        type: array
+        description: Taxes that are applied to this invoice and calculated based on the related product `taxCategoryId`.
```
Comment (noon-dawg) [inline openapi/components/schemas/Invoices/TaxTypes/rebilly-avalara.yaml:12]: nitpick: for consistency with the description in `/schemas/Invoices/TaxTypes/rebilly-taxjar.yaml`
```suggestion
        description: Taxes that are applied to this invoice and calculated based on the `taxCategoryId` of the related product.
```

File: openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml [hunk 19]
```diff
@@ -0,0 +1,24 @@
+allOf:
```
Comment (adamaltman) [inline openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml:1]: This folder structure doesn't make sense.

**Review by adamaltman — DISMISSED** [review]
Do not make new folders. I'll adjust the guidance. But making folders for one file hurts my soul.

File: openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml (hunk 19 above)
Comment (subenson) [reply openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml:1]: Shall I move every specific Credential.yaml out of it's subdirectory into `/Credentials/`, while keeping the selected ones in the screenshot below? 

image

Comment (subenson) [reply openapi/components/schemas/Credentials/AvalaraCredentials/AvalaraCredential.yaml:1]: I just restructured the Avalara Credential as this PR is blocking. On approval, we can apply the changes to the other credentials as well.

**Comment by subenson** [comment]
@Rebilly/php @Rebilly/docs @Rebilly/js please review

**Review by noon-dawg — COMMENTED** [review]
Looks good to me!

**Review by mikerebilly — DISMISSED** [review]

**Comment by subenson** [comment]
@Rebilly/php @Rebilly/js please review

**Review by lexerom — APPROVED** [review]

**Review by ninjasimon — APPROVED** [review]

**Comment by adamaltman** [comment]
FYI this violates the style guide. Tags should be alphabetical. But even more important, there should not be an avalara tag. This is bad organization -- imagine if we made a tag for each gateway integration?

**Pull request merged**